### PR TITLE
Switch from tinyxml -> tinyxml2

### LINKED ISSUE #137: Build fails with undefined reference to TiXmlElement::SetAttribute
Ubuntu focal / Clang 11 / libc++
Building the ros2/urdfdom ros2 branch

```
--- stderr: urdfdom                                                                                                                                   
ld.lld: error: ../lib/liburdfdom_model.so.1.0: undefined reference to TiXmlElement::SetAttribute(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [urdf_parser/CMakeFiles/check_urdf.dir/build.make:88: bin/check_urdf] Error 1
make[1]: *** [CMakeFiles/Makefile2:294: urdf_parser/CMakeFiles/check_urdf.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
ld.lld: error: ../lib/liburdfdom_model.so.1.0: undefined reference to TiXmlElement::SetAttribute(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
ld.lld: error: ../lib/liburdfdom_model.so.1.0: undefined reference to TiXmlElement::SetAttribute(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [urdf_parser/CMakeFiles/urdf_mem_test.dir/build.make:87: bin/urdf_mem_test] Error 1
make[1]: *** [CMakeFiles/Makefile2:564: urdf_parser/CMakeFiles/urdf_mem_test.dir/all] Error 2
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [urdf_parser/CMakeFiles/urdf_to_graphiz.dir/build.make:87: bin/urdf_to_graphiz] Error 1
make[1]: *** [CMakeFiles/Makefile2:348: urdf_parser/CMakeFiles/urdf_to_graphiz.dir/all] Error 2
ld.lld: error: ../../lib/liburdfdom_model.so.1.0: undefined reference to TiXmlElement::SetAttribute(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [urdf_parser/test/CMakeFiles/urdf_unit_test.dir/build.make:89: bin/urdf_unit_test] Error 1
make[1]: *** [CMakeFiles/Makefile2:1106: urdf_parser/test/CMakeFiles/urdf_unit_test.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
---
Failed   <<< urdfdom	[ Exited with code 2 ]
```

### PR DESCRIPTION
**Pull request by rotu**
Fix https://github.com/ros/urdfdom/issues/137

### CODE REVIEW

File: urdf_parser/src/xml_helpers.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
+/*********************************************************************
```
Comment (clalancette) [inline urdf_parser/src/xml_helpers.h:1]: Rename this file to `xml_helpers.hpp`, please.

File: urdf_parser/src/xml_helpers.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+#ifndef ROS2_MASTER_XML_HELPERS_HPP
+#define ROS2_MASTER_XML_HELPERS_HPP
```
Comment (clalancette) [inline urdf_parser/src/xml_helpers.h:36]:
```suggestion
#ifndef URDF_PARSER__XML_HELPERS_HPP
#define URDF_PARSER__XML_HELPERS_HPP
```

File: urdf_parser/include/urdf_parser/urdf_parser.h [hunk 3]
```diff
@@ -48,18 +48,19 @@
 
 #include "exportdecl.h"
 
+using TiXmlDocument = tinyxml2::XMLDocument;
+using TiXmlElement = tinyxml2::XMLElement;
```
Comment (clalancette) [inline urdf_parser/include/urdf_parser/urdf_parser.h:52]: I'd really rather not pretend that tinyxml2 APIs are the same as the tinyxml ones; this is changing the API below, but in a sneaky way.  Let's remove this, deprecate the tinyxml APIs, and also add in the tinyxml2 APIs.

File: urdf_parser/src/joint.cpp [hunk 4]
```diff
@@ -639,46 +642,42 @@ bool exportPose(Pose &pose, TiXmlElement* xml);
 
 bool exportJointDynamics(JointDynamics &jd, TiXmlElement* xml)
 {
-  TiXmlElement *dynamics_xml = new TiXmlElement("dynamics");
-  dynamics_xml->SetAttribute("damping", urdf_export_helpers::values2str(jd.damping) );
-  dynamics_xml->SetAttribute("friction", urdf_export_helpers::values2str(jd.friction) );
-  xml->LinkEndChild(dynamics_xml);
+  auto * dynamics_xml = xmlAppendChild(xml, "dynamics");
```
Comment (clalancette) [inline urdf_parser/src/joint.cpp:645]: I really prefer not to use `auto` unless it is blindingly obvious from this line of code what the type is.  It doesn't seem to be the case here, so I'd much prefer using `tinyxml2::XMLElement*` here.  The same goes for most uses of `auto` below.

File: urdf_parser/src/joint.cpp [hunk 5]
```diff
@@ -639,46 +642,42 @@ bool exportPose(Pose &pose, TiXmlElement* xml);
...
 {
-  TiXmlElement *limit_xml = new TiXmlElement("limit");
-  limit_xml->SetAttribute("effort", urdf_export_helpers::values2str(jl.effort) );
-  limit_xml->SetAttribute("velocity", urdf_export_helpers::values2str(jl.velocity) );
-  limit_xml->SetAttribute("lower", urdf_export_helpers::values2str(jl.lower) );
-  limit_xml->SetAttribute("upper", urdf_export_helpers::values2str(jl.upper) );
-  xml->LinkEndChild(limit_xml);
+  auto * limit_xml = xmlAppendChild(xml, "limit");
+  limit_xml->SetAttribute("effort", jl.effort);
+  limit_xml->SetAttribute("velocity", jl.velocity );
```
Comment (clalancette) [inline urdf_parser/src/joint.cpp:655]: Nit: no space before the closing parentheses.  There are more instances of this below.

**Review by clalancette — CHANGES_REQUESTED** [review]
I've pointed out some things to fix in here.  Most important looks to be the changes to the APIs; we'll have to tick-tock those.

**Comment by rotu** [comment]
@clalancette, feel free to take this over.

**Pull request closed without merging**